### ISSUE TITLE
Feature: ssh gateway hosts

### ISSUE THREAD

**Issue by zswanson** [njpanderson/push#7]
Add configuration ability to tunnel sftp through an intermediate ssh host (a gateway host)

**Comment by njpanderson** [comment]
Hi there @zswanson - Not quite sure of the mechanics of this but will look and see if the underlying SSH libs support it and let you know.

**Comment by njpanderson** [comment]
So to clarify broadly what you're asking for, this is the ability to define two SSH servers in your SFTP config and for Push to use the first as an intermediary to the second?

I've had a little look into this and from what I can tell, the feature you're talking about is also referred to as "Connection hopping" by the underlying SSH lib. There's more information here:

https://github.com/mscdex/ssh2#connection-hopping
http://www.onlamp.com/pub/a/onlamp/excerpt/ssh_11/index1.html

If this is indeed what you're looking for, I'll need to set up an SSH environment in order to replicate this, and it looks like the solution relies on the (fairly common) "nc" utility on *nix OSes. Your environment will no doubt be much different to mine, so if there's anything you can tell me about it that might help me to look into things I'd be very grateful.

**Comment by zswanson** [comment]
Yes typically this is for cases where a network is separated by a 'bastion'
host that you ssh hop through. Not sure about using nc, usually this is
achieved by creating an ssh to the intermediate, then tunneling through
that connection to the remote host.

I had checked around the various VScode extensions and nobody seems to
offer this, but gateway/bastion hosts are quite common in enterprise/cloud
work.

**Comment by njpanderson** [comment]
I'm working on this. It's curently in branch [`try/gateway-ssh`](https://github.com/njpanderson/push/tree/try/gateway-ssh) and you're welcome to give it a bash if you know how to use `vsce` to package up vscode extensions for local testing. (More information about vsce here: https://code.visualstudio.com/docs/extensions/publish-extension — if if you don't fancy that, you can just use the debug mode in vscode with this repo checked out.

**Comment by zswanson** [comment]
Fantastic I'll take a look! This will be a big time saver for our work flow.

**Comment by njpanderson** [comment]
No problem - let me know if you manage to test this. I don't have access to a real environment for this, so I've been faking it with a UNIX SSH gateway. Things might go a little differently in windows...

**Comment by njpanderson** [comment]
Hey @zswanson - Did you manage to test this? Reluctant to release this without someone else having a go with it first.

**Comment by zswanson** [comment]
Sorry, was out for a week and had to get an environment setup to actually
run this in debug mode. I think I've got the debug extension running now, I
just need to copy in the new settings and give it a shot.

**Comment by njpanderson** [comment]
That's totally fine! Just let me know if you do manage to test it, otherwise I'll give it a few more run throughs in different envs and see how it goes before releasing it.

**Comment by zswanson** [comment]
Ran it this morning, works like a charm!

**Comment by zswanson** [comment]
I have 2 different bastions, both linux, that serve a set of linux clusters
behind the bastion. I used ssh keys to access the bastion and user/password
for the hosts in the dmz (which are temporary VMs for testing). Tested on
both bastions to a couple of different hosts on the backend, worked fine.

**Comment by njpanderson** [comment]
That's great, thanks a lot for confirming. It's pegged for release v0.3 which I'll try to get out in the next couple of weekends.
